fix error message for XDR readers

orbeckst opened the pull request:
Fixes #

Changes made in this Pull Request:
- error message for not being able to write lockfile did not insert the filename into message: fixed by using proper f string
- changed all .format to modern f string in XDR.py
- add explicit test for offsets_filename()
- add additional pytest.warns() to tests to catch warnings (reduces warnings output and checks that all expected warnings are raised with proper messages)



PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
:books: Documentation preview :books:: https://mdanalysis--4231.org.readthedocs.build/en/4231/

<!-- readthedocs-preview mdanalysis end -->

orbeckst committed:
fix error message for XDR readers

- error message for not being able to write lockfile did not insert the filename into
  message: fixed by using proper f string
- changed all .format to modern f string in XDR.py
- add explicit test for offsets_filename()
- add additional pytest.warns() to tests to catch warnings (reduces warnings output and
  checks that all expected warnings are raised with proper messages)

orbeckst commented:
FYI, this is what the (small) error looks like in practice
```
>>> u = mda.Universe("md-01.tpr", "md-01.part0001.xtc")
~/miniconda3/envs/mda25/lib/python3.10/site-packages/MDAnalysis/coordinates/XDR.py:203: UserWarning: Cannot write lock/offset file in same location as {self.filename}. Using slow offset calculation.
  warnings.warn(f"Cannot write lock/offset file in same location as "
~/miniconda3/envs/mda25/lib/python3.10/site-packages/MDAnalysis/coordinates/XDR.py:259: UserWarning: Couldn't save offsets because: [Errno 13] Permission denied: '.md-01.part0001.xtc_offsets.npz'
  warnings.warn("Couldn't save offsets because: {}".format(e))
```

Note how `{self.filename}` is not replaced.

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -46,6 +47,14 @@
 from MDAnalysis.coordinates import XDR
 from MDAnalysisTests.util import get_userid
 
+@pytest.mark.parametrize("filename,kwargs,reference", [
```
```suggestion

@pytest.mark.parametrize("filename,kwargs,reference", [
```

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -46,6 +47,14 @@
+])
+def test_offsets_filename(filename, kwargs, reference):
+    fn = XDR.offsets_filename(filename, **kwargs)
+    assert fn == reference
```
```suggestion
    assert fn == reference

```

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -755,25 +764,34 @@ def test_nonexistent_offsets_file(self, traj):
-            trajectory._load_offsets()
+            with pytest.warns(UserWarning) as record:
+                trajectory._load_offsets()
+            assert len(record) == 2
```
checking the length of the record is very flaky, can we instead just check that the two warnings you're looking for are in there? "X in Y" type of thing

orbeckst replied:
ok, added in [4f18afb](https://github.com/MDAnalysis/mdanalysis/pull/4231/commits/4f18afbb8e423386daf600cd6b13e56158e91eeb)

IAlibay reviewed (changes requested):
Couple of small things, otherwise lgtm

orbeckst committed:
more robust checking of multiple warnings

orbeckst committed:
Apply suggestions from code review

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

IAlibay reviewed (approved):
cheers, lgtm!

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -755,25 +766,43 @@ def test_nonexistent_offsets_file(self, traj):
+        assert saved_offsets == False
 
-    def test_nonexistent_offsets_file(self, traj):
+    def test_corrupted_offsets_file(self, traj):
```
btw, this shadowed the preceeding test; renaming increased our tests

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -755,25 +766,43 @@ def test_nonexistent_offsets_file(self, traj):
+            assert len(record) >= len(warning_messages)
+            found_messages = [str(r.message) for r in record]
+            # find at least 2 occurences (>= instead of == for robustness)
+            assert sum([ref in msg for msg in found_messages
```
I looked through the other tests in the file and found the following approach that looks a bit cleaner where we stack pytest.warns:
```python
       with (pytest.warns(UserWarning, match="Couldn't save offsets") and
              pytest.warns(UserWarning, match="Cannot write")):
            self._reader(filename)
```

I'll rewrite to use that idiom.

orbeckst committed:
simplified testing for multiple warnings

orbeckst committed:
Merge branch 'fix-xdr-errormessages' of github.com:MDAnalysis/mdanalysis into fix-xdr-errormessages

orbeckst committed:
more explicit warnings testing

- reduces warnings in test_xdr down to 2 (both related to del)
- test for empty selection and missing elements warnings
- rewriting more old-school parameter interpolation to f strings

orbeckst commented:
Thanks for the quick review @IAlibay . I apologize, I tacked on a few little things that I couldn't let go when I looked at the code again.

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -542,7 +553,10 @@ class _GromacsWriterIssue117(object):
     def universe(self):
-        return mda.Universe(PRMncdf, NCDF)
+        with pytest.warns(UserWarning,
+                          match="ATOMIC_NUMBER record not found"):
```
I'm so confused, why are testing a NetCDF file warning here? Should we use pytest.filterwarning if it's not something we need to explicitly test?

orbeckst replied:
My thinking was that it filters the warning and alerts us if anything is changing. Given how poor our handling of warning in tests is generally, that seemed reasonable.

However, I also see that it can be confusing to test something not directly related to the code in question (here: xdr).

What's the appropriate `filterwarnings` to use here?

IAlibay replied:
> it filters the warning and alerts us if anything is changing

We should already be testing that warning in the netcdf tests (I'm pretty sure we do because I tried to bring testing to 100% via test_netcdf.py last I touched it). I'll leave it to your decision here (feel free to dismiss my block, everything else looks good).

> What's the appropriate filterwarnings to use here?

Adding the following to the tests calling the fixture I think should do the trick:
`@pytest.mark.filterwarnings("ignore: ATOMIC_NUMBER record not found")`

orbeckst replied:
Thanks. Changed.

RMeli commented on package/MDAnalysis/coordinates/XDR.py:
```diff
@@ -220,10 +219,10 @@ def _load_offsets(self):
-                              "Consider setting 'refresh_offsets=True' "
-                              "when loading your Universe.".format(fname))
+                warnings.warn(f"Reading offsets from {fname} failed, "
+                              f"reading offsets from trajectory instead.\n"
+                              f"Consider setting 'refresh_offsets=True' "
+                              f"when loading your Universe.")
```
```suggestion
                              "reading offsets from trajectory instead.\n"
                              "Consider setting 'refresh_offsets=True' "
                              "when loading your Universe.")
```

RMeli replied:
No need for f-strings here? (Unless there it's some black/formatting thing I'm not aware of...) IIRC some linters might complain.

RMeli commented on testsuite/MDAnalysisTests/coordinates/test_xdr.py:
```diff
@@ -551,18 +565,19 @@ def test_write_trajectory(self, universe, tmpdir):
-                "between original and written "
-                "trajectory at frame %d (orig) vs %d "
-                "(written)" % (orig_ts.frame, written_ts.frame))
+                err_msg=(f"coordinate mismatch between original and written "
```
```suggestion
                err_msg=("coordinate mismatch between original and written "
```

orbeckst committed:
only use f-strings where used

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

IAlibay committed:
ignore warnings not germane to the XDR tests

orbeckst committed:
Merge branch 'fix-xdr-errormessages' of github.com:MDAnalysis/mdanalysis into fix-xdr-errormessages

orbeckst commented:
One of the runners timed out so I restarted it.

orbeckst commented:
@RMeli or @IAlibay could one of you please volunteer to captain the PR and eventually merge it? I can do it if you prefer but it's better style (in my opinion) to not be the one merging your own PR (unless it's urgent....).

I think it's good for the project when there's visibly someone in charge of a PR.

IAlibay commented:
lgtm!
